## Loops
In this course we’ll learn about structures that let you repeat Python code over and over. You will learn how to understand loops and write them to solve your own problems

## Introduction
Suppose we want to print() each item from a list of dog_breeds. We would need to use the following code snippet:

In [1]:
dog_breeds = ['french_bulldog', 'dalmatian', 'shihtzu', 'poodle', 'collie']
 
print(dog_breeds[0])
print(dog_breeds[1])
print(dog_breeds[2])
print(dog_breeds[3])
print(dog_breeds[4])

french_bulldog
dalmatian
shihtzu
poodle
collie


This seems inefficient. Luckily, Python (and most other programming languages) gives us an easier way of using, or iterating through, every item in a list. We can use loops! A loop is a way of repeating a set of code many times.

In this lesson, we’ll be learning about:

- Loops that let us move through each item in a list, called for loops
- Loops that keep going until we tell them to stop, called while loops
- Loops that create new lists, called list comprehensions

### Applications
Paste the following code into the cell below.

In [2]:
#This will print each breed in dog_breeds.
for breed in dog_breeds:
    print(breed)

french_bulldog
dalmatian
shihtzu
poodle
collie


In [3]:
#This will print each breed in dog_breeds.
for x in dog_breeds:
    print(x)

french_bulldog
dalmatian
shihtzu
poodle
collie


#### Possible Question

What kind of loop operations are available in Python?

#### Possible answer

There are three types of loops in Python. The for loop iterates over a list. The while loop executes a block of code repeatedly while the provided expression evaluates to True. A list comprehension is a specialized way to construct a new list using a built-in loop over another list.

## Create a For Loop
In the previous exercise, we saw that we can print each item in a list using a for loop. A for loop lets us perform an action on each item in a list. Using each element of a list is known as iterating.

This loop prints each breed in dog_breeds:

In [ ]:
dog_breeds = ['french_bulldog', 'dalmation', 'shihtzu', 'poodle', 'collie']
for breed in dog_breeds:
    print(breed)

The general way of writing a for loop is:

In [4]:
# for <temporary variable> in <list variable>:
#    <action>

In our dog breeds example, breed was the temporary variable, dog_breeds was the list variable, and print(breed) was the action performed on every item in the list.

Our temporary variable can be named whatever we want and does not need to be defined beforehand. Each of the following code snippets does the exact same thing as our example:

In our dog breeds example, breed was the temporary variable, dog_breeds was the list variable, and print(breed) was the action performed on every item in the list.

Our temporary variable can be named whatever we want and does not need to be defined beforehand. Each of the following code snippets does the exact same thing as our example:

In [5]:
for i in dog_breeds:
    print(i)

french_bulldog
dalmatian
shihtzu
poodle
collie


In [6]:
for dog in dog_breeds:
    print(dog)

french_bulldog
dalmatian
shihtzu
poodle
collie


Notice that in all of these examples the print statement is indented. Everything in the same level of indentation after the for loop declaration is included in the for loop, and run every iteration.

If we forget to indent, we’ll get an IndentationError.

### Applications

1.
Run the code. You should get an IndentationError because the print(game) line is not indented.

2.
Indent line 6 so that you don’t get an IndentationError when you run the code.

In [7]:
board_games = ['Settlers of Catan', 'Carcassone', 'Power Grid', 'Agricola', 'Scrabble']

sport_games = ['football', 'football - American', 'hockey', 'baseball', 'cricket']

for game in board_games:
print(game)

IndentationError: expected an indented block (<ipython-input-7-e8dabf1b79d3>, line 6)

3.
Write a loop that prints each sport in sport_games.

In [8]:
#code here

sport_games = ['football', 'football - American', 'hockey', 'baseball', 'cricket']

for game in sport_games:
    print(game)


football
football - American
hockey
baseball
cricket


In [9]:
sport_games = ['football', 'football - American', 'hockey', 'baseball', 'cricket']

for game in sport_games:
    print(game*2)


footballfootball
football - Americanfootball - American
hockeyhockey
baseballbaseball
cricketcricket


### Possible Question
In this exercise 3, the code for the for loop has to be indented. Why does it have to be?

### Possible Answer
Indentation in Python is used to identify lines of code which are executed together. The block of code that is part of for loop has to be indented so that Python knows what to execute for each iteration of the loop. The point at which the indentation ends identifies the code that is NOT part of the for loop.

In the following code example, the print() call which outputs “In loop” will be repeated as part of the for loop but the next immediate line containing the print() which outputs “Not in loop” will execute only once.

In [10]:
#Example
for x in range(5):
    print("In loop")
print("Not in loop")

In loop
In loop
In loop
In loop
In loop
Not in loop


## Using Range in Loops
Previously, we iterated through an existing list.

Often we won’t be iterating through a specific list, we’ll just want to do a certain action multiple times. For example, if we wanted to print out a "WARNING!" message three times, we would want to say something like:

In [ ]:
# for i in <a list of length 3>:
#    print("WARNING!")

Notice that we need to iterate through a list of length 3, but we don’t care what’s in the list. To create these lists of length n, we can use the range function. range takes in a number n as input, and returns a list from 0 to n-1. For example:

In [14]:
zero_thru_five = range(6)
print(list(zero_thru_five))
# zero_thru_five is now [0, 1, 2, 3, 4, 5]

zero_thru_one = range(2)
print(list(zero_thru_one))
# zero_thru_one is now [0, 1]

[0, 1, 2, 3, 4, 5]
[0, 1]


So, an easy way to accomplish our "WARNING!" example would be:


In [15]:
for i in range(3):
    print("WARNING!")

WARNING!
WARNING!
WARNING!


### Applications
1.
Use the range function in a for loop to print out promise 5 times.

In [16]:
promise = "I will not chew gum in class"

# code here
for a in range(5):
    print(promise)


I will not chew gum in class
I will not chew gum in class
I will not chew gum in class
I will not chew gum in class
I will not chew gum in class


2. Use the range function in a for loop to print out the square of each number from 1 thru 10

In [20]:
# code here

for i in range(1,11):
    print(i**2)

1
4
9
16
25
36
49
64
81
100


In [21]:
for i in range(10):
    print((i+1)**2)

1
4
9
16
25
36
49
64
81
100


In [22]:
for i in range(1, 11):
    print(i*i)


1
4
9
16
25
36
49
64
81
100


## Infinite Loops
We’ve iterated through lists that have a discrete beginning and end. However, let’s consider this example:

In [ ]:
my_favorite_numbers = [4, 8, 15, 16, 42]
 
for number in my_favorite_numbers:
    my_favorite_numbers.append(1)

What happens here? Every time we enter the loop, we add a 1 to the end of the list that we are iterating through. As a result, we never make it to the end of the list! It keeps growing!

A loop that never terminates is called an infinite loop. These are very dangerous for your code!

A program that hits an infinite loop often becomes completely unusable. The best course of action is to never write an infinite loop.

Note: If you accidentally stumble into an infinite loop while developing on your own machine, you can end the loop by using control + c to terminate the program. If you’re writing code in our online editor, you’ll need to refresh the page to get out of an infinite loop!

### Applications
1.
Suppose we have two lists of students, students_period_A and students_period_B. We want to combine all students into students_period_B.

Write a for loop that goes through each student in students_period_A and adds it to the end of students_period_B.

In [24]:
students_period_A = ["Alex", "Briana", "Cheri", "Daniele"]
students_period_B = ["Dora", "Minerva", "Alexa", "Obie"]
#code here

for i in students_period_A:
    students_period_B.append(i)
    
print(students_period_B)

['Dora', 'Minerva', 'Alexa', 'Obie', 'Alex', 'Briana', 'Cheri', 'Daniele']


2.
Inside the for loop, after appending student to students_period_B, print student.

In [25]:
#code here
print(students_period_B)

['Dora', 'Minerva', 'Alexa', 'Obie', 'Alex', 'Briana', 'Cheri', 'Daniele']


3.
Let’s suppose you made a typo in the body of your for loop. Oh no! We will go through a few steps to see what happens when you end up with an infinite loop and how you fix it:

Inside the for loop, change the object of the append statement from students_period_B to students_period_A.

Run this code. What do you notice happens? Over the run button, notice the loading circle is continuing without anything happening. This is an infinite loop! To end this program we must refresh the page. (Note: The reason this loop is infinite is that we’re adding each student in students_period_A to students_period_A which would create a never-ending list of all the student names.)

After refreshing your page, you should see that the run button is back. Now get rid of the change we made that caused an infinite loop by changing the object of the append statement back to student_period_B.

Run your code again. This time your loop should not be infinite, and it should run to completion.



In [26]:
# code here
for i in students_period_A:
    students_period_B.append(i)

4.
Great job! We have now seen what happens when you run an infinite loop and how we can fix them! Revel in the safety of your code, and how it doesn’t crash your browser. Run it one final time to make sure it is working as expected.

In [ ]:
# code here


## Break
We often want to use a for loop to search through a list for some value:

In [27]:
items_on_sale = ["blue_shirt", "striped_socks", "knit_dress", "red_headband", "dinosaur_onesie"]
 
# we want to check if the item with ID "knit_dress" is on sale:
for item in items_on_sale:
    if item == "knit_dress":
        print("Knit Dress is on sale!")

Knit Dress is on sale!


This code goes through each item in items_on_sale and checks for a match. After we find that "knit_dress" is in the list items_on_sale, we don’t need to go through the rest of the items_on_sale list. Since it’s only 5 elements long, iterating through the entire list is not a big deal in this case. But what if items_on_sale had 1000 items after "knit_dress"? What if it had 100,000 items after "knit_dress"?

You can stop a for loop from inside the loop by using break. When the program hits a break statement, control returns to the code outside of the for loop. For example:

In [28]:
items_on_sale = ["blue_shirt", "striped_socks", "knit_dress", "red_headband", "dinosaur_onesie"]
 
print("Checking the sale list!")
for item in items_on_sale:
    print(item)
    if item == "knit_dress":
        break
print("End of search!")

Checking the sale list!
blue_shirt
striped_socks
knit_dress
End of search!


This would produce the output:

Checking the sale list!
blue_shirt
striped_socks
knit_dress
End of search!
We didn’t need to check "red_headband" or "dinosaur_onesie" at all!

#### Applications
1.
You have a list of dog breeds you can adopt, dog_breeds_available_for_adoption. Using a for loop, iterate through the dog_breeds_available_for_adoption list and print out each dog breed.

In [30]:
dog_breeds_available_for_adoption = ['french_bulldog', 'dalmatian', 'shihtzu', 'poodle', 'collie']
dog_breed_I_want = 'dalmatian'
# code here
for i in dog_breeds_available_for_adoption:
    print(i)

french_bulldog
dalmatian
shihtzu
poodle
collie


2.
Inside your for loop, after you print each dog breed, check if it is equal to dog_breed_I_want. If so, print "They have the dog I want!"

In [34]:
dog_breeds_available_for_adoption = ['french_bulldog', 'dalmatian', 'shihtzu', 'poodle', 'collie']
dog_breed_I_want = 'dalmatian'
# code here
for i in dog_breeds_available_for_adoption:
    print(i)
    if i == dog_breed_I_want:
        print("They have the dog I want")
        break

french_bulldog
dalmatian
They have the dog I want


3.
Add a break statement when your loop has found dog_breed_I_want, so that the rest of the list does not need to be checked.

In [35]:
dog_breeds_available_for_adoption = ['french_bulldog', 'dalmatian', 'shihtzu', 'poodle', 'collie']
dog_breed_I_want = 'dalmatian'
# code here
for i in dog_breeds_available_for_adoption:
    print(i)
    if i == dog_breed_I_want:
        print("They have the dog I want")
        break

french_bulldog
dalmatian
They have the dog I want


## Continue
When we’re iterating through lists, we may want to skip some values. Let’s say we want to print out all of the numbers in a list, unless they’re negative. We can use continue to move to the next i in the list:

In [37]:
big_number_list = [1, 2, -1, 4, -5, 5, 2, -9, 0]
 
for i in big_number_list:
    if i < 0:
        continue
    print(i)

1
2
4
5
2
0


Every time there was a negative number, the continue keyword moved the index to the next value in the list, without executing the code in the rest of the for loop.

#### Applications

1.
Your computer is the doorman at a bar in a country where the drinking age is 21.

Loop through the ages list. If an entry is less than 21, skip it and move to the next entry. Otherwise, print the age.

In [45]:
ages = [21, 35, 12, 38, 34, 26, 21, 19, 67, 41, 17]

#code here
for age in ages:
    if age < 21 :
        continue
    print(age)
        

21
35
38
34
26
21
67
41


## While Loops
We now have seen and used a lot of examples of for loops. There is another type of loop we can also use, called a while loop. The while loop performs a set of code until some condition is reached.

While loops can be used to iterate through lists, just like for loops:

In [46]:
dog_breeds = ['bulldog', 'dalmation', 'shihtzu', 'poodle', 'collie']
 
index = 0
while index < len(dog_breeds):
    print(dog_breeds[index])
    index += 1

bulldog
dalmation
shihtzu
poodle
collie


Every time the condition of the while loop (in this case, index < len(dog_breeds)) is satisfied, the code inside the while loop runs.

While loops can be useful when you don’t know how many iterations it will take to satisfy a condition.

#### Applications
1.
You are adding students to a Poetry class, the size of which is capped at 6. While the length of the students_in_poetry list is less than 6, use .pop() to take a student off the all_students list and add it to the students_in_poetry list.

In [47]:
all_students = ["Alex", "Briana", "Cheri", "Daniele", "Dora", "Minerva", "Alexa", "Obie", "Arius", "Loki"]
students_in_poetry = []

#code here
while len(students_in_poetry)<6:
    student = all_students.pop()
    students_in_poetry.append(student)

print(all_students)
print(students_in_poetry)
    

['Alex', 'Briana', 'Cheri', 'Daniele']
['Loki', 'Arius', 'Obie', 'Alexa', 'Minerva', 'Dora']


In [49]:
all_students = ["Alex", "Briana", "Cheri", "Daniele", "Dora", "Minerva", "Alexa", "Obie", "Arius", "Loki"]
print(all_students.pop())
print(all_students)

Loki
['Alex', 'Briana', 'Cheri', 'Daniele', 'Dora', 'Minerva', 'Alexa', 'Obie', 'Arius']


## Nested Loops
We have seen how we can go through the elements of a list. What if we have a list made up of multiple lists? How can we loop through all of the individual elements?

Suppose we are in charge of a science class, that is split into three project teams:

In [51]:
project_teams = [["Ava", "Samantha", "James"], 
                 ["Lucille", "Zed"], 
                 ["Edgar", "Gabriel"]]

If we want to go through each student, we have to put one loop inside another:

In [52]:
for team in project_teams:
    for student in team:
        print(student)

Ava
Samantha
James
Lucille
Zed
Edgar
Gabriel


In [53]:
for team in project_teams:
    print(team)

['Ava', 'Samantha', 'James']
['Lucille', 'Zed']
['Edgar', 'Gabriel']


#### Applications


In [55]:
sales_data = [[12, 17, 22], [2, 10, 3], [5, 12, 13]]



1.
We have provided the list sales_data that shows the numbers of different flavors of ice cream sold at three different locations of the fictional shop, Gilbert and Ilbert’s Scoop Shop. We want to sum up the total number of scoops sold. Start by defining a variable scoops_sold and set it to zero.

In [57]:
scoops_sold = 0

for location in sales_data:
    for i in location:
        scoops_sold += i
        print(scoops_sold)
    print(location)

12
29
51
[12, 17, 22]
53
63
66
[2, 10, 3]
71
83
96
[5, 12, 13]


2.
Go through the sales_data list. Call each inner list location, and print out each location list.

3.
Within the sales_data loop, go through each location list and add the element to your scoops_sold variable.

By the end, you should have the sum of every number in the sales_data nested list.

4. Print out the value of scoops_sold.



## List Comprehensions
Let’s say we have scraped a certain website and gotten these words:

In [59]:
words = ["@coolguy35",
         "#nofilter", 
         "@kewldawg54", 
         "reply", 
         "timestamp", 
         "@matchamom", 
         "follow", 
         "#updog"]

We want to make a new list, called usernames, that has all of the strings in words with an '@' as the first character. We know we can do this with a for loop:

In [61]:
usernames = []
 
for word in words:
    if word[0] == '@':
        usernames.append(word)
        
print(usernames)

['@coolguy35', '@kewldawg54', '@matchamom']


First, we created a new empty list, usernames, and as we looped through the words list, we added every word that matched our criterion. Now, the usernames list looks like this:

In [ ]:
print(usernames)
["@coolguy35", "@kewldawg54", "@matchamom"]

Python has a convenient shorthand to create lists like this with one line:

In [63]:
usernames = [word for word in words if word[0] == '@']

print(usernames)

['@coolguy35', '@kewldawg54', '@matchamom']


In [71]:
words = ["@coolguy35",
         "#nofilter", 
         "@kewldawg54", 
         "reply", 
         "timestamp", 
         "@matchamom", 
         "follow", 
         "#updog"]

usernames = [word for word in words if word[0] == '@']

print(usernames)

['@coolguy35', '@kewldawg54', '@matchamom']


This is called a list comprehension. It will produce the same output as the for loop did:

In [ ]:
["@coolguy35", "@kewldawg54", "@matchamom"]

This list comprehension:

1. Takes an element in words
2. Assigns that element to a variable called word
3. Checks if word[0] == '@', and if so, it adds word to the new list, usernames. If not, nothing happens.
4. Repeats steps 1-3 for all of the strings in words

</Note>: if we hadn’t done any checking (let’s say we had omitted if word[0] == '@'), the new list would be just a copy of words:




In [65]:
usernames = [word for word in words]
print(usernames)
#usernames is now ["@coolguy35", "#nofilter", "@kewldawg54", "reply", "timestamp", "@matchamom", "follow", "#updog"]

['@coolguy35', '#nofilter', '@kewldawg54', 'reply', 'timestamp', 'matchamom@', 'follow', '#updog']


#### Applications
1.
We have defined a list heights of visitors to a theme park. In order to ride the Topsy Turvy Tumbletron roller coaster, you need to be above 161 centimeters. Using a list comprehension, create a new list called can_ride_coaster that has every element from heights that is greater than 161.

In [67]:
heights = [161, 164, 156, 144, 158, 170, 163, 163, 157]

can_ride_coaster = [height for height in heights if height > 161]

print(can_ride_coaster)


[164, 170, 163, 163]


In [68]:
heights = [161, 164, 156, 144, 158, 170, 163, 163, 157]

can_ride_coaster = []

for h in heights:
    if h > 161:
        can_ride_coaster.append(h)

print(can_ride_coaster)

2.
Print can_ride_coaster.

In [69]:
print(can_ride_coaster)

[164, 170, 163, 163]


#### Possible Question
When using a list comprehension with an if, is it possible to have an else clause?

#### Possible Answer
Yes, an else clause can be used with an if in a list comprehension. The following code example shows the use of an else in a simple list comprehension. The if/else is placed in front of the for component of the list comprehension.


In [70]:
divbythree = [  "Yes" if number % 3 == 0 else "No" for number in range(1,20)]
print(divbythree)

['No', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'No', 'Yes', 'No']


## More List Comprehensions
Let’s say we’re working with the usernames list from the last exercise:

In [72]:
print(usernames)
["@coolguy35", "@kewldawg54", "@matchamom"]

['@coolguy35', '@kewldawg54', '@matchamom']


['@coolguy35', '@kewldawg54', '@matchamom']

We want to create a new list with the string " please follow me!" added to the end of each username. We want to call this new list messages. We can use a list comprehension to make this list with one line:

In [74]:
messages = [user + " please follow me!" for user in usernames]
print(messages)

['@coolguy35 please follow me!', '@kewldawg54 please follow me!', '@matchamom please follow me!']


This list comprehension:

1. Takes a string in usernames
2. Assigns that string to a variable called user
3. Adds “ please follow me!” to user
4. Appends that concatenation to the new list called messages
5. Repeats steps 1-4 for all of the strings in usernames

Now, messages contains these values:

["@coolguy35 please follow me!", "@kewldawg54 please follow me!", "@matchamom please follow me!"]

Being able to create lists with modified values is especially useful when working with numbers. Let’s say we have this list:

In [75]:
my_upvotes = [192, 34, 22, 175, 75, 101, 97]

We want to add 100 to each value. We can accomplish this goal in one line:

In [76]:
updated_upvotes = [vote_value + 100 for vote_value in my_upvotes]
print(updated_upvotes)

[292, 134, 122, 275, 175, 201, 197]


This list comprehension:

1. Takes a number in my_upvotes
2. Assigns that number to a variable called vote_value
3. Adds 100 to vote_value
4. Appends that sum to the new list updated_upvotes
5. Repeats steps 1-4 for all of the numbers in my_upvotes

Now, updated_upvotes contains these values:
    
[292, 134, 122, 275, 175, 201, 197]

#### Applications

1.
We have provided a list of temperatures in celsius. Using a list comprehension, create a new list called fahrenheit that converts each element in the celsius list to fahrenheit.

*Note: * To convert, use the formula:

temperature_in_fahrenheit = temperature_in_celsius * 9/5 + 32

2.
Print fahrenheit.

In [77]:
celsius = [0, 10, 15, 32, -5, 27, 3]



In [83]:
# You can even use functions embeded in the new list you created, uncomment the codes below

def temperature_in_fahrenheit(temperature_in_celsius):
    return temperature_in_celsius * 9/5 + 32

fahrenheit = [temperature_in_fahrenheit(temp) for temp in celsius]

print(fahrenheit)

[32.0, 50.0, 59.0, 89.6, 23.0, 80.6, 37.4]


## Review
Good job! In this lesson, you learned

1. how to write a for loop
2. how to use range in a loop
3. what infinite loops are and how to avoid them
4. how to skip values in a loop
5. how to write a while loop
6. how to make lists with one line

Let’s get some more practice with these concepts!

#### Applications
1.
Create a list called single_digits that consists of the numbers 0-9 (inclusive).

2.
Create a for loop that goes through single_digits and prints out each one.

3.
Before the loop, create a list called squares. Assign it to be an empty list to begin with.

4.
Inside the loop that iterates through single_digits, append the squared value of each element of single_digits to the list squares. You can do this before or after printing the element.

5.
After the for loop, print out squares

6.
Create the list cubes using a list comprehension on the single_digits list. Each element of cubes should be an element of single_digits taken to the third power.

7.
Print cubes.

Good job!

In [86]:
single_digits = list(range(10))

squares = []
for digit in single_digits:
    squares.append(digit**2)
    
print(squares)

cubes =[]
for digit in single_digits:
    cubes.append(digit**3)

print(cubes)



[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 8, 27, 64, 125, 216, 343, 512, 729]


In [85]:
squares = [digit ** 2 for digit in single_digits]
print(squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


## Looping Techniques
This is an additional document that you might want to utilize from.

https://docs.python.org/3/tutorial/datastructures.html#looping-techniques

# Quiz

1. What would be the output of the following code:

#drink_choices = ["coffee", "tea", "water", "juice", "soda"]
#for drink in drink_choices:
#print(drink)

A.[“coffee”, “tea”, “water”, “juice”, “soda”]


B.drink


C.coffee


D. 
coffee

tea

water

juice

soda

In [88]:
drink_choices = ["coffee", "tea", "water", "juice", "soda"]
for drink in drink_choices:
    print(drink)

coffee
tea
water
juice
soda


2. What would be the output of the following code:

In [89]:
numbers = [1, 1, 2, 3]
for number in numbers:
    if number % 2 == 0:
        break
    print(number)

1
1


A.
1
1

B.
2

C.
1
1
2
3

D.
1
1
3

3. What would be the output of the following code:



In [90]:
numbers = [1, 1, 2, 3, 4, 5, 6, 3, 5]
for number in numbers:
    if number % 2 == 0:
        continue
    print(number)

1
1
3
5
3
5


A.
2

B.
1
1
3

C.
1
1
2

D.
1
1

1. What would be the output of the following code:



In [ ]:
#for i in range(3):
#    print(5)

In [94]:
for i in range(3):
    print(5)

5
5
5


5. Fill in the blank with the appropriate while condition in order to print the numbers 1 through 10 in order:



In [96]:
i = 1
while i <= 10:
    print(i)
    i += 1

1
2
3
4
5
6
7
8
9
10


A. while i > 10:


B. while i == range(11):


C. while i < 10:


D. while i <= 10:

6. Which of these list comprehensions will create a list equal to desired_list?


In [100]:
my_list = [5, 10, -2, 8, 20]
#desired_list = [10, 8, 20]
print([i for i in my_list if i > 5])

[10, 8, 20]


A. [i for i in my_list if i > 5]

B. [i + 5 for i in my_list]

C. [i for i in my_list]

D. [i for i in my_list if i > 10]

7. Which of these list comprehensions will create a list equal to desired_list?

In [31]:
#desired_list = [-1, 0, 1, 2, 3]

A. [i for i in range(5)]


B. range(5)


C. [i-1 for i in range(5)]

8.What would be the output of the following code:

In [103]:
for i in range(3):
    print(i)

0
1
2


A.
i
i
i
i

B.
1
2
3

C.
i
i
i

D.
0
1
2

9. What would be the output of the following code:

In [104]:
numbers = [2, 4, 6, 8]
for number in numbers:
    print("hello!")

hello!
hello!
hello!
hello!


In [107]:
numbers = [2, 4, 6, 8]
for number in numbers:
    print(*"hello!" * number) 

# This might be an interesting one to understand it

h e l l o ! h e l l o !
h e l l o ! h e l l o ! h e l l o ! h e l l o !
h e l l o ! h e l l o ! h e l l o ! h e l l o ! h e l l o ! h e l l o !
h e l l o ! h e l l o ! h e l l o ! h e l l o ! h e l l o ! h e l l o ! h e l l o ! h e l l o !


A. hello!


B.
hello!
hello!
hello!
hello!


C.. 2 hello! 4 hello! 6 hello! 8 hello!


D. 2 4 6 8

In [37]:
#numbers = [2, 4, 6, 8]
#for number in numbers:
#    print("hello!")